Lab5/ScannerIntegration/Parser.py

```diff
@@ -0,0 +1,233 @@
+                for i in range(0, lengthProduction - 1):
+                    self.tree[vectorIndex[i]].sibling = vectorIndex[i + 1]
+
+            else:
+                self.tree[index].father = father
+                father = -1
+
+        filename = self.filename.split(".")[0] + ".out"
+        file = open(filename, "w")
+        file.write("index | value | father | sibling\n")
+
+        for index in range(0, len(work)):
+            file.write(str(index) + " " + str(self.tree[index]) + "\n")
+            print(index, " ", str(self.tree[index]))
+
+        file.close()
+
+    def getProductionOffset(self, index):
+        production = self.grammar.getProductions()[self.work[index][0]][self.work[index][1]]
+        lengthOfProduction = len(production)
+        offset = lengthOfProduction
+        for i in range(1, lengthOfProduction + 1):
+            if type(self.work[index + i]) == tuple:
+                offset += self.getProductionOffset(index + i)
+        return offset
+
+    def run(self, w):
+        self.word = w
+        self.state = "q"
+        self.index = 0
+        self.work = []
+        self.input = [self.grammar.getStartSymbol()]
+        while self.state != "t" and self.state != "e":
+            if self.state == "q":
+                # Done condition
+                if len(self.input) == 0 and self.index == len(w):
+                    self.state = "t"
+                else:
+                    # expand condition, top of beta is non terminal
+                    if self.input[0] in self.grammar.getNonTerminals():
+                        self.expand()
+                    # advance condition, top of beta is terminal and it matches the index of the w
+                    elif self.input[0] in self.grammar.getTerminals() and self.checkWordLength(w):
+                        self.advance()
+                    # Woopsie -> we have a momentary insuccess
+                    else:
+                        self.momentaryInsuccess()
+            else:
+                if self.state == "b":
+                    # back
+                    if self.work[len(self.work) - 1] in self.grammar.getTerminals():
+                        self.back()
+                    # anotherTry
+                    else:
+                        self.anotherTry()
+        if self.state == "e":
+            print("ERROR")
+        else:
+            print("Sequence accepted")
+            print(self.work, self.input, self.index)
+
+
+class Node:
+    # this is a node in the parse tree, an arbitrary tree using the father sibling representation
+    # all nodes are connected with only 2 links - the father and the right sibling
+    # these nodes will be stored in a python list
+    def __init__(self, value):
+        self.father = -1
+        self.sibling = -1
+        self.value = value
+        self.production = -1
+
+    def __str__(self):
+        return str(self.value) + " " + str(self.father) + " " + str(self.sibling)
+
+

```

Comment:
There should be a separate file that uses Parser, Grammar and others. 
So it can be more stateless.